File: study/lua/src/app/ui/component/CommonSelectNumNode.lua
local CommonIconNameNode = class("CommonIconNameNode")

local EXPORTED_METHODS = {
	"updateAmount",
	"getAmount",
	"setTextDesc",
    "setMaxLimit",
    "setMaxLimitEx",
	"setCallBack",
	"setAmount",
	"showButtonMax",
    "getMaxLimit",
    "updateIncrement",
}

--
function CommonIconNameNode:ctor()
	self._target = nil
	self._button = nil
	self._seleclCallBack = nil

	self._composeCount = 0
	self._maxLimit = 999
	
end

function CommonIconNameNode:getMaxLimit()
	return self._maxLimit
end
--
function CommonIconNameNode:_init()
	self._textBatchUse = ccui.Helper:seekNodeByName(self._target,"Text_batch_use")
	self._btnSub10 = ccui.Helper:seekNodeByName(self._target,"Button_sub10")
	self._btnSub1 = ccui.Helper:seekNodeByName(self._target,"Button_sub1")
	self._btnAdd10 = ccui.Helper:seekNodeByName(self._target,"Button_add10")
	self._btnMax = ccui.Helper:seekNodeByName(self._target,"Button_max")
	self._btnAdd1 = ccui.Helper:seekNodeByName(self._target,"Button_add1")
	self._textAmount = ccui.Helper:seekNodeByName(self._target, "Text_amount")

	self:updateAmount(0)

	local function callBack(isAdd)
		if self._seleclCallBack then
			self._seleclCallBack(self._composeCount, isAdd)
		end
	end
	self._target:updateButton("Button_sub10", function()
        self:updateAmount(-10)
		callBack(false)
    end, -500)
    self._target:updateButton("Button_sub1", function()
        self:updateAmount(-1)
		callBack(false)
    end, -500)
    self._target:updateButton("Button_add10", function()
        self:updateAmount(10)
		callBack(true)
    end, -500)

	self._target:updateButton("Button_max", function()
        self:updateAmount(self._maxLimit)
		callBack(true)
    end, -500)

    self._target:updateButton("Button_add1", function()
        self:updateAmount(1)
		callBack(true)
    end, -500)

	self:showButtonMax(false)
end

--
function CommonIconNameNode:bind(target)
	self._target = target
    self:_init()
    cc.setmethods(target, self, EXPORTED_METHODS)
end

--
function CommonIconNameNode:unbind(target)
    cc.unsetmethods(target, EXPORTED_METHODS)
end

--
function CommonIconNameNode:setCallBack(func)
	if func and type(func) == "function" then
		self._seleclCallBack = func 
	end
end

function CommonIconNameNode:setMaxLimit(max)
	if max > 999 then
		max = 999
	end
	self._maxLimit = max or 999
end

function CommonIconNameNode:setMaxLimitEx(max)
	self._maxLimit = max
end

function CommonIconNameNode:updateAmount(num)
	--dump( self._composeCount )
	local addNum = self._composeCount + num

	self:setAmount(addNum)
end

function CommonIconNameNode:getAmount()
	return self._composeCount
end

function CommonIconNameNode:setAmount(num)
	self._composeCount = num
	
	if self._composeCount >= self._maxLimit then
		self._composeCount = self._maxLimit
	end
	if self._composeCount <= 0 then
		self._composeCount = 1
	end
	self._textAmount:setString(""..self._composeCount)
end


function CommonIconNameNode:setTextDesc(desc)
	if self._textBatchUse then
		self._textBatchUse:setString(desc)
		self._textBatchUse:setVisible(true)
	end
end

function CommonIconNameNode:showButtonMax(needShow)
	if needShow then
		self._btnMax:setVisible(true)
		self._btnAdd10:setVisible(false)
	else
		self._btnMax:setVisible(false)
		self._btnAdd10:setVisible(true)
	end
end

function CommonIconNameNode:_updateAmount( ... )
    -- body
    local function callBack(isAdd)
        if self._seleclCallBack then
            self._seleclCallBack(self._composeCount, isAdd)
        end
    end
    self._target:updateButton("Button_sub10", function()
        self:updateAmount(-100)
        callBack(false)
    end, -500)

    self._target:updateButton("Button_add10", function()
        self:updateAmount(100)
        callBack(true)
    end, -500)

    self._target:updateButton("Button_max", function()
        self:updateAmount(self._maxLimit)
        callBack(true)
    end, -500)
end

function CommonIconNameNode:updateIncrement(num)
    -- body
    local iconSub = self._btnSub10:getChildByName("Image_1")
    local iconAdd = self._btnAdd10:getChildByName("Image_1")
    if iconSub and iconAdd then
        if rawequal(num, 100) then
            iconSub:loadTexture(Path.getUICommon("img_com_btn_minus03"))
            iconAdd:loadTexture(Path.getUICommon("img_com_btn_minus04"))
            self:_updateAmount()
        end
    end
end

return CommonIconNameNode

File: study/ts/script/ui/component/CommonSelectNumNode.ts
import { Path } from "../../utils/Path";
import UIHelper from "../../utils/UIHelper";

const { ccclass, property } = cc._decorator;

@ccclass
export default class CommonSelectNumNode extends cc.Component {
    @property({
        type: cc.Node,
        visible: true
    })
    _panel_addInfo: cc.Node = null;
    @property({
        type: cc.Label,
        visible: true
    })
    _textBatchUse: cc.Label = null;
    @property({
        type: cc.Button,
        visible: true
    })
    _btnSub10: cc.Button = null;
    @property({
        type: cc.Button,
        visible: true
    })
    _btnSub1: cc.Button = null;
    @property({
        type: cc.Sprite,
        visible: true
    })
    _image_amount_bg: cc.Sprite = null;
    @property({
        type: cc.Label,
        visible: true
    })
    _textAmount: cc.Label = null;
    @property({
        type: cc.Button,
        visible: true
    })
    _btnAdd10: cc.Button = null;
    @property({
        type: cc.Button,
        visible: true
    })
    _btnMax: cc.Button = null;
    @property({
        type: cc.Sprite,
        visible: true
    })
    _image_1_0: cc.Sprite = null;
    @property({
        type: cc.Button,
        visible: true
    })
    _btnAdd1: cc.Button = null;
    @property({
        type: cc.Sprite,
        visible: true
    })
    _popupCommonItemUse_1: cc.Sprite = null;
    _maxLimit = 999;
    _seleclCallBack: any;
    _composeCount: number = 0;

    getMaxLimit() {
        return this._maxLimit;
    }
    onLoad() {
        this.updateAmount(0);
        var clickEventHandler = new cc.Component.EventHandler();
        clickEventHandler.target = this.node; // 这个 node 节点是你的事件处理代码组件所属的节点
        clickEventHandler.component = "CommonSelectNumNode";// 这个是代码文件名
        clickEventHandler.handler = "_onSub10";
        this._btnSub10.clickEvents.push(clickEventHandler)

        clickEventHandler = new cc.Component.EventHandler();
        clickEventHandler.target = this.node; // 这个 node 节点是你的事件处理代码组件所属的节点
        clickEventHandler.component = "CommonSelectNumNode";// 这个是代码文件名
        clickEventHandler.handler = "_onSub1";
        this._btnSub1.clickEvents.push(clickEventHandler)

        clickEventHandler = new cc.Component.EventHandler();
        clickEventHandler.target = this.node; // 这个 node 节点是你的事件处理代码组件所属的节点
        clickEventHandler.component = "CommonSelectNumNode";// 这个是代码文件名
        clickEventHandler.handler = "_onAdd10";
        this._btnAdd10.clickEvents.push(clickEventHandler)

        clickEventHandler = new cc.Component.EventHandler();
        clickEventHandler.target = this.node; // 这个 node 节点是你的事件处理代码组件所属的节点
        clickEventHandler.component = "CommonSelectNumNode";// 这个是代码文件名
        clickEventHandler.handler = "_onAdd1";
        this._btnAdd1.clickEvents.push(clickEventHandler)

        clickEventHandler = new cc.Component.EventHandler();
        clickEventHandler.target = this.node; // 这个 node 节点是你的事件处理代码组件所属的节点
        clickEventHandler.component = "CommonSelectNumNode";// 这个是代码文件名
        clickEventHandler.handler = "_onAddMax";
        this._btnMax.clickEvents.push(clickEventHandler)

        this.showButtonMax(false);
    }
    _onSub10() {
        this._callBack(false, 10);
    }
    _onSub1() {
        this._callBack(false, 1);
    }
    _onAdd10() {
        this._callBack(true, 10);
    }
    _onAdd1() {
        this._callBack(true, 1);
    }
    _onAddMax() {
        this._callBack(true, this._maxLimit);
    }
    setMaxLimitEx(max) {
        this._maxLimit = max;
    }

    _callBack(isAdd, num) {
        this.updateAmount(isAdd ? num : -num);
        if (this._seleclCallBack) {
            this._seleclCallBack(this._composeCount, isAdd);
        }
    }

    setCallBack(func) {
        if (func && typeof (func) == 'function') {
            this._seleclCallBack = func;
        }
    }
    setMaxLimit(max) {
        if (max > 999) {
            max = 999;
        }
        this._maxLimit = max || 999;
    }
    updateAmount(num) {
        var addNum = this._composeCount + num;
        this.setAmount(addNum);
    }
    getAmount() {
        return this._composeCount;
    }
    setAmount(num) {
        this._composeCount = num;
        if (this._composeCount >= this._maxLimit) {
            this._composeCount = this._maxLimit;
        }
        if (this._composeCount <= 0) {
            this._composeCount = 1;
        }
        this._textAmount.string = ('' + this._composeCount);
    }
    setTextDesc(desc) {
        if (this._textBatchUse) {
            this._textBatchUse.string = (desc);
            this._textBatchUse.node.active = (true);
        }
    }
    showButtonMax(needShow) {
        if (needShow) {
            this._btnMax.node.active = (true);
            this._btnAdd10.node.active = (false);
        } else {
            this._btnMax.node.active = (false);
            this._btnAdd10.node.active = (true);
        }
    }
    _updateAmount() {
        let callBack = function (isAdd) {
            if (this._seleclCallBack) {
                this._seleclCallBack(this._composeCount, isAdd);
            }
        }.bind(this);
        this._btnSub10.clickEvents = [];
        this._btnAdd10.clickEvents = [];
        this._btnMax.clickEvents = [];

        this._btnSub10.node.on(cc.Node.EventType.TOUCH_START, function () {
            this.updateAmount(-1000);
            callBack(false);
        }.bind(this));
        this._btnAdd10.node.on(cc.Node.EventType.TOUCH_START, function () {
            this.updateAmount(1000);
            callBack(true);
        }.bind(this));
        this._btnMax.node.on(cc.Node.EventType.TOUCH_START, function () {
            this.updateAmount(this._maxLimit);
            callBack(true);
        }.bind(this));
    }
    updateIncrement(num) {
        var iconSub = this._btnSub10.node.getChildByName('Image_1').getComponent(cc.Sprite);
        var iconAdd = this._btnAdd10.node.getChildByName('Image_1').getComponent(cc.Sprite);
        if (iconSub && iconAdd) {
            if (num ==100) {
                UIHelper.loadTexture(iconSub,Path.getUICommon('img_com_btn_minus03'));
                UIHelper.loadTexture(iconAdd,Path.getUICommon('img_com_btn_minus04'));
                this._updateAmount();
            }
            else if(num==1000)
            {
                UIHelper.loadTexture(iconSub,Path.getUICommon('img_com_btn_minus06'));
                UIHelper.loadTexture(iconAdd,Path.getUICommon('img_com_btn_minus05'));
                this._updateAmount();
            }
        }
    }
}